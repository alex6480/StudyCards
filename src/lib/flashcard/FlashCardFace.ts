import { ContentState, convertToRaw } from "draft-js";

export type FlashCardFaceId = "back" | "front";

export enum FlashCardFaceType {
    None,
    Image,
    RichText,
}

export interface IFlashCardFace {
    id: FlashCardFaceId;
    setId: string;
    cardId: string;
    type: FlashCardFaceType;
    richTextContent: ContentState;
}

export interface IRichTextFlashCardFace extends IFlashCardFace {
    type: FlashCardFaceType.RichText;
}

export interface INoFlashCardFace extends IFlashCardFace {
    type: FlashCardFaceType.None;
}

export interface IImageFlashCardFace extends IFlashCardFace {
    type: FlashCardFaceType.Image;
}

export type ExportFlashCardFace = ExportRichTextFlashCardFace | ExportNoFlashCardFace | ExportImageFlashCardFace;

export class ExportRichTextFlashCardFace {
    public readonly id: string;
    public readonly type: FlashCardFaceType.RichText;
    public readonly richTextContent: string;

    constructor(face: IRichTextFlashCardFace) {
        this.id = face.id;
        this.type = face.type;
        const rawContent = face.richTextContent != null ? convertToRaw(face.richTextContent) : null;
        this.richTextContent = JSON.stringify(rawContent);
        // this.richTextContent = draftToMarkdown(rawContent, {})
    }
}

export class ExportNoFlashCardFace {
    public readonly id: string;
    public readonly type: FlashCardFaceType.None;

    constructor(face: INoFlashCardFace) {
        this.id = face.id;
        this.type = face.type;
    }
}

export class ExportImageFlashCardFace {
    public readonly id: string;
    public readonly type: FlashCardFaceType.Image;

    constructor(face: IImageFlashCardFace) {
        this.id = face.id;
        this.type = face.type;
    }
}
