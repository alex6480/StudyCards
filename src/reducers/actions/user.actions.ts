import { createAction } from ".";
import { IUser } from "../../lib/User";

export const LOG_OUT = "LOG_OUT";
export const AUTHENTICATE = "AUTHENTICATE";

export const userActions = {
    logOut: () => createAction(LOG_OUT),
    authenticate: (user: IUser) => createAction(AUTHENTICATE, { user }),
};
