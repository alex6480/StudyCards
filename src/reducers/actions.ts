import { Dispatch } from "redux";
import IFlashCard, { IFlashCardMeta } from "../lib/flashcard/flashcard";
import { IFlashCardFace } from "../lib/flashcard/FlashCardFace";
import IFlashCardSet, { IFlashCardSetMeta } from "../lib/flashcard/FlashCardSet";
import { ICardStudyData, ISetStudyData } from "../lib/flashcard/StudyData";

/*
    Boilerplate stuff used to get typesafety within Redux
*/
export interface IAction<T extends string> {
    type: T;
}
export interface IActionWithPayload<T extends string, P> extends IAction<T> {
    payload: P;
}

export function createAction<T extends string>(type: T): IAction<T>;
export function createAction<T extends string, P>(type: T, payload: P): IActionWithPayload<T, P>;
export function createAction<T extends string, P>(type: T, payload?: P) {
    return payload === undefined ? { type } : { type, payload };
}

type FunctionType = (...args: any[]) => any;
interface IActionsCreatorMapObject {
    [actionCreator: string]: FunctionType;
}

export type ActionsUnion<A extends IActionsCreatorMapObject> = ReturnType<A[keyof A]>;

export const RESET_SESSION_STUDY_DATA = "reset session study data";
export const UPDATE_CARD_STUDY_DATA = "update card study data";
export const SWAP_CARD_FACES = "swap card faces";

// Remote actions
export const DELETE_CARD_BEGIN = "delete card begin";
export const DELETE_CARD_COMPLETE = "delete card complete";

export const SAVE_CARD_FACE_BEGIN = "save card face begin";
export const SAVE_CARD_FACE_COMPLETE = "save card face complete";

export const ADD_NEW_SET_BEGIN = "add new set begin";
export const ADD_NEW_SET_COMPLETE = "add new set complete";

export const ADD_NEW_CARD_BEGIN = "add new card begin";
export const ADD_NEW_CARD_COMPLETE = "add new card complete";
export const ADD_NEW_CARD_ERROR = "add new card error";

export const LOAD_SET_META_ALL_BEGIN = "load meta data for all sets begin";
export const LOAD_SET_META_ALL_COMPLETE = "load meta data for all sets complete";
export const LOAD_SET_META_ALL_ERROR = "load meta data for all sets error";

export const LOAD_SET_STUDY_DATA_BEGIN = "load set study data begin";
export const LOAD_SET_STUDY_DATA_COMPLETE = "load set study data complete";

export const LOAD_CARDS_BEGIN = "load cards begin";
export const LOAD_CARDS_COMPLETE = "load cards complete";

export const SAVE_CARD_META_BEGIN = "save card meta begin";
export const SAVE_CARD_META_COMPLETE = "save card meta complete";

export const SAVE_SET_META_BEGIN = "save set meta begin";
export const SAVE_SET_META_COMPLETE = "save set meta complete";

export const Action = {
    loadSetMetaAllBegin: () => createAction(LOAD_SET_META_ALL_BEGIN),
    loadSetMetaAllComplete: (setMeta: {[id: string]: IFlashCardSetMeta}) =>
        createAction(LOAD_SET_META_ALL_COMPLETE, setMeta),
    loadSetMetaAllError: (message: string) => createAction(LOAD_SET_META_ALL_ERROR, { message }),

    loadSetStudyDataBegin: (setId: string) => createAction(LOAD_SET_STUDY_DATA_BEGIN, { setId }),
    loadSetStudyDataComplete: (result: ISetStudyData) =>
        createAction(LOAD_SET_STUDY_DATA_COMPLETE, { setId: result.setId, result }),

    loadCardsBegin: (setId: string, cardIds: string[]) => createAction(LOAD_CARDS_BEGIN, { setId, cardIds }),
    loadCardsComplete: (setId: string, cards: {[id: string]: IFlashCard }) =>
        createAction(LOAD_CARDS_COMPLETE, { setId, cards }),

    addNewCardBegin: (cardId: string, setId: string, afterCardId?: string) =>
        createAction(ADD_NEW_CARD_BEGIN, { cardId, setId, afterCardId }),
    addNewCardComplete: (setId: string, cardId: string) =>
        createAction(ADD_NEW_CARD_COMPLETE, { setId, cardId }),

    addSetBegin: (setId: string, set: Partial<IFlashCardSet>) =>
        createAction(ADD_NEW_SET_BEGIN, { setId, set }),
    addSetComplete: (setId: string) => createAction(ADD_NEW_SET_COMPLETE, { setId }),

    saveCardFaceBegin: (setId: string, cardId: string, face: IFlashCardFace) =>
        createAction(SAVE_CARD_FACE_BEGIN, { setId, cardId, face }),
    saveCardFaceComplete: (setId: string, cardId: string) =>
        createAction(SAVE_CARD_FACE_COMPLETE, { setId, cardId }),

    saveCardMetaBegin: (setId: string, cardId: string, cardMeta: Partial<IFlashCardMeta>) =>
        createAction(SAVE_CARD_META_BEGIN, { setId, cardId, cardMeta }),
    saveCardMetaComplete: (setId: string, cardId: string, cardMeta: IFlashCardMeta) =>
        createAction(SAVE_CARD_META_COMPLETE, { setId, cardId, cardMeta }),

    saveSetMetaBegin: (setMeta: Partial<IFlashCardSetMeta>) =>
        createAction(SAVE_SET_META_BEGIN, { setId: setMeta.id, setMeta }),
    saveSetMetaComplete: (setId: string) =>
        createAction(SAVE_SET_META_COMPLETE, { setId }),

    deleteCardBegin: (setId: string, cardId: string) => createAction(DELETE_CARD_BEGIN, { cardId, setId }),
    deleteCardComplete: (setId: string, cardId: string) => createAction(DELETE_CARD_COMPLETE, { cardId, setId }),

    updateCardStudyData: (studyData: ICardStudyData) => createAction(UPDATE_CARD_STUDY_DATA, studyData),
    swapCardFaces: (setId: string, cardId: string) => createAction(SWAP_CARD_FACES, { cardId, setId }),

    /**
     * Resets the parts of the studydata that are temporary for a single study session
     */
    resetSessionStudyData: () => createAction(RESET_SESSION_STUDY_DATA),
};

export type Action = ActionsUnion<typeof Action>;
