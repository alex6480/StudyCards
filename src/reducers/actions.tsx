import FlashCard from "../lib/flashcard/flashcard";
import FlashCardSet from "../lib/flashcard/FlashCardSet";
import { FlashCardFace } from "../lib/flashcard/FlashCardFace";

/*
    Boilerplate stuff used to get typesafety within Redux
*/
export interface Action<T extends string> {
    type: T;
}
export interface ActionWithPayload<T extends string, P> extends Action<T> {
    payload: P;
}

export function createAction<T extends string> (type: T): Action<T>
export function createAction<T extends string, P> (type: T, payload: P): ActionWithPayload<T, P>
export function createAction<T extends string, P> (type: T, payload?: P) {
    return payload == undefined ? { type } : { type, payload };
}

type FunctionType = (...args: any[]) => any;
type ActionsCreatorMapObject = { [actionCreator: string]: FunctionType };

export type ActionsUnion<A extends ActionsCreatorMapObject> = ReturnType<A[keyof A]>;

export const ADD_NEW_CARD = "add new card";
export const UPDATE_SET_NAME = "update set name";
export const DELETE_CARD = "delete card";
export const UPDATE_CARD_FACE = "update card face";
export const ADD_NEW_SET = "add new set";

export const Actions = {
    addNewCard: (setId: string) => createAction(ADD_NEW_CARD, { setId: setId }),
    updateSetName: (set: FlashCardSet, newName: string) => createAction(UPDATE_SET_NAME, { set: set, name: newName }),
    deleteCard: (card: FlashCard) => createAction(DELETE_CARD, card),
    updateCardFace: (cardId: string, face: FlashCardFace) => createAction(UPDATE_CARD_FACE, {
        cardId: cardId,
        face: face
    }),
    addSet: (set?: FlashCardSet) => createAction(ADD_NEW_SET, { set: set })
}

export type Actions = ActionsUnion<typeof Actions>;