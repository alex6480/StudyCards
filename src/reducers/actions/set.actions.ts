import { createAction } from ".";
import { IFlashCardSetMeta } from "../../lib/flashcard/FlashCardSet";

export const LIST_SETS_BEGIN = "LIST_SETS_BEGIN";
export const LIST_SETS_COMPLETE = "LIST_SETS_COMPLETE";
export const LIST_SETS_ERROR = "LIST_SETS_ERROR";

export const setActions = {
    loadSetMetaAllBegin: () => createAction(LIST_SETS_BEGIN),
    loadSetMetaAllComplete: (setMeta: {[id: string]: IFlashCardSetMeta}) =>
        createAction(LIST_SETS_COMPLETE, setMeta),
    loadSetMetaAllError: () => createAction(LIST_SETS_ERROR, { }),
};
