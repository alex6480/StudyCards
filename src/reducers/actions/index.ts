import { Dispatch } from "redux";
import { ThunkAction } from "redux-thunk";
import { IAppState } from "..";
import IFlashCard, { IFlashCardMeta } from "../../lib/flashcard/flashcard";
import { IFlashCardFace } from "../../lib/flashcard/FlashCardFace";
import IFlashCardSet, { IFlashCardFilter, IFlashCardSetMeta } from "../../lib/flashcard/FlashCardSet";
import { ICardStudyData, ISetStudyData } from "../../lib/flashcard/StudyData";
import { IStudyState } from "../../lib/flashcard/StudyState";
import { CardEvaluation } from "../../lib/study";
import { userActions } from "./user.actions";

/*
    Boilerplate stuff used to get typesafety within Redux
*/
export interface IAction<T extends string> {
    type: T;
}
export interface IActionWithPayload<T extends string, P> extends IAction<T> {
    payload: P;
}

export function createAction<T extends string>(type: T): IAction<T>;
export function createAction<T extends string, P>(type: T, payload: P): IActionWithPayload<T, P>;
export function createAction<T extends string, P>(type: T, payload?: P) {
    return payload === undefined ? { type } : { type, payload };
}

type FunctionType = (...args: any[]) => any;
interface IActionsCreatorMapObject {
    [actionCreator: string]: FunctionType;
}

export type ActionsUnion<A extends IActionsCreatorMapObject> = ReturnType<A[keyof A]>;
export type TAction<R, E> = ThunkAction<R, IAppState, E, Action>;

export const SWAP_CARD_FACES = "swap card faces";

// Remote actions
export const DELETE_CARD_BEGIN = "delete card begin";
export const DELETE_CARD_COMPLETE = "delete card complete";

export const SAVE_CARD_FACE_BEGIN = "save card face begin";
export const SAVE_CARD_FACE_COMPLETE = "save card face complete";

export const ADD_NEW_SET_BEGIN = "add new set begin";
export const ADD_NEW_SET_COMPLETE = "add new set complete";

export const ADD_NEW_CARD_BEGIN = "add new card begin";
export const ADD_NEW_CARD_COMPLETE = "add new card complete";
export const ADD_NEW_CARD_ERROR = "add new card error";

export const LOAD_SET_META_ALL_BEGIN = "load meta data for all sets begin";
export const LOAD_SET_META_ALL_COMPLETE = "load meta data for all sets complete";
export const LOAD_SET_META_ALL_ERROR = "load meta data for all sets error";

export const LOAD_CARDS_BEGIN = "load cards begin";
export const LOAD_CARDS_COMPLETE = "load cards complete";

export const SAVE_CARD_META_BEGIN = "save card meta begin";
export const SAVE_CARD_META_COMPLETE = "save card meta complete";

export const SAVE_SET_META_BEGIN = "save set meta begin";
export const SAVE_SET_META_COMPLETE = "save set meta complete";

export const SET_FILTER_CARDS_BEGIN = "set filter cards begin";
export const SET_FILTER_CARDS_COMPLETE = "set filter cards complete";

export const SET_STUDY_STATE_BEGIN = "set study state begin";
export const SET_STUDY_STATE_COMPLETE = "set study state complete";

export const EVALUATE_CARD_BEGIN = "evaluate card begin";
export const EVALUATE_CARD_COMPLETE = "evaluate card complete";

export const Action = {
    ...userActions,

    loadSetMetaAllBegin: () => createAction(LOAD_SET_META_ALL_BEGIN),
    loadSetMetaAllComplete: (setMeta: {[id: string]: IFlashCardSetMeta}) =>
        createAction(LOAD_SET_META_ALL_COMPLETE, setMeta),
    loadSetMetaAllError: (message: string) => createAction(LOAD_SET_META_ALL_ERROR, { message }),

    loadCardsBegin: (setId: string, cardIds: string[]) => createAction(LOAD_CARDS_BEGIN, { setId, cardIds }),
    loadCardsComplete: (setId: string, cards: {[id: string]: IFlashCard }) =>
        createAction(LOAD_CARDS_COMPLETE, { setId, cards }),

    addNewCardBegin: (cardId: string, setId: string, afterCard?: IFlashCard) =>
        createAction(ADD_NEW_CARD_BEGIN, { cardId, setId, afterCard }),
    addNewCardComplete: (setId: string, cardId: string) =>
        createAction(ADD_NEW_CARD_COMPLETE, { setId, cardId }),

    addSetBegin: (setId: string, set: Partial<IFlashCardSet>) =>
        createAction(ADD_NEW_SET_BEGIN, { setId, set }),
    addSetComplete: (setId: string) => createAction(ADD_NEW_SET_COMPLETE, { setId }),

    saveCardFaceBegin: (setId: string, cardId: string, face: IFlashCardFace) =>
        createAction(SAVE_CARD_FACE_BEGIN, { setId, cardId, face }),
    saveCardFaceComplete: (setId: string, cardId: string) =>
        createAction(SAVE_CARD_FACE_COMPLETE, { setId, cardId }),

    saveCardMetaBegin: (setId: string, cardId: string, cardMeta: Partial<IFlashCardMeta>) =>
        createAction(SAVE_CARD_META_BEGIN, { setId, cardId, cardMeta }),
    saveCardMetaComplete: (setId: string, cardId: string, cardMeta: IFlashCardMeta) =>
        createAction(SAVE_CARD_META_COMPLETE, { setId, cardId, cardMeta }),

    saveSetMetaBegin: (setMeta: Partial<IFlashCardSetMeta>) =>
        createAction(SAVE_SET_META_BEGIN, { setId: setMeta.id, setMeta }),
    saveSetMetaComplete: (setId: string) =>
        createAction(SAVE_SET_META_COMPLETE, { setId }),

    deleteCardBegin: (setId: string, cardId: string) => createAction(DELETE_CARD_BEGIN, { cardId, setId }),
    deleteCardComplete: (setId: string, cardId: string) => createAction(DELETE_CARD_COMPLETE, { cardId, setId }),

    swapCardFaces: (setId: string, cardId: string) => createAction(SWAP_CARD_FACES, { cardId, setId }),

    filterCardsBegin: (setId: string, filter: IFlashCardFilter) =>
        createAction(SET_FILTER_CARDS_BEGIN, { setId, filter} ),
    filterCardsComplete: (setId: string, filter: IFlashCardFilter, result: string[]) =>
        createAction(SET_FILTER_CARDS_COMPLETE, { setId, filter, result }),

    setStudyStateBegin: (studyState?: Partial<IStudyState>) => createAction(SET_STUDY_STATE_BEGIN, { studyState }),
    setStudyStateComplete: (studyState: Partial<IStudyState>) => createAction(SET_STUDY_STATE_COMPLETE, { studyState }),

    evaluateCardBegin: (setId: string, cardId: string, evaluation: CardEvaluation) =>
        createAction(EVALUATE_CARD_BEGIN, { setId, cardId, evaluation }),
    evaluateCardComplete: (setId: string, cardId: string, evaluation: CardEvaluation,
                           redrawTime: Date | null, nextCardId: string, studyState: IStudyState) =>
        createAction(EVALUATE_CARD_COMPLETE, { setId, cardId, evaluation, redrawTime, nextCardId, studyState }),
};

export type Action = ActionsUnion<typeof Action>;
