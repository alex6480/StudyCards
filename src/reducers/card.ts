import { ContentState } from "draft-js";
import IFlashCard from "../lib/flashcard/flashcard";
import { FlashCardFaceType } from "../lib/flashcard/FlashCardFace";
import IRemote from "../lib/remote";
import * as Utils from "../lib/utils";
import * as fromActions from "./actions";

export const initialCard: IFlashCard = {
    id: "",
    setId: "",
    faces: {
        front: {
            id: "front",
            cardId: "",
            setId: "",
            type: FlashCardFaceType.RichText,
            richTextContent: ContentState.createFromText(""),
        },
        back: {
            id: "back",
            cardId: "",
            setId: "",
            type: FlashCardFaceType.RichText,
            richTextContent: ContentState.createFromText(""),
        },
    },
};
export const initialState = { isFetching: true, lastUpdated: Date.now(), value: initialCard };

function value(state: Partial<IFlashCard> = initialState.value,
               cardId: string, action: fromActions.Action): IFlashCard {
    switch (action.type) {
        default:
            return {
                id: cardId,
                setId: state.setId as string,
                faces: state.faces !== undefined ? state.faces : initialCard.faces,
            };
    }
}

export default function card(state: IRemote<Partial<IFlashCard>> = initialState,
                             cardId: string,
                             action: fromActions.Action): IRemote<IFlashCard> {
    switch (action.type) {
        case fromActions.LOAD_CARDS_BEGIN:
            return {
                ...state,
                value: value(state.value, cardId, action),
                isFetching: true,
            };
        case fromActions.LOAD_CARDS_COMPLETE:
            return {
                ...state,
                value: action.payload.cards[cardId],
                isFetching: false,
            };
        default:
            return {
                ...state,
                value: value(state.value, cardId, action),
            };
    }
}
