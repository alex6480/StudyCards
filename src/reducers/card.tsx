import { ContentState } from "draft-js";
import IFlashCard from "../lib/flashcard/flashcard";
import { FlashCardFaceType } from "../lib/flashcard/FlashCardFace";
import * as Utils from "../lib/utils";
import * as fromActions from "./actions";

export const initialState: IFlashCard = {
    id: "",
    setId: "",
    faces: {
        front: {
            id: "front",
            cardId: "",
            setId: "",
            type: FlashCardFaceType.RichText,
            richTextContent: ContentState.createFromText(""),
        },
        back: {
            id: "back",
            cardId: "",
            setId: "",
            type: FlashCardFaceType.RichText,
            richTextContent: ContentState.createFromText(""),
        },
    },
};

export default function card(state: IFlashCard = initialState, action: fromActions.Actions): IFlashCard {
    switch (action.type) {
        case fromActions.ADD_NEW_CARD:
            if (state.id === initialState.id) {
                return { ...state,
                    id: Utils.guid(),
                    setId: action.payload.setId };
            }
            break;
        case fromActions.UPDATE_CARD_FACE:
            if (action.payload.cardId === state.id) {
                return { ...state,
                    faces: {...state.faces,
                        [action.payload.face.id]: action.payload.face,
                    },
                };
            }
            break;
        default:
            return state;
    }
    return state;
}
