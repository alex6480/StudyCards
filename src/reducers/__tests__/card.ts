import { ContentState } from "draft-js";
import IFlashCard from "../../lib/flashcard/flashcard";
import { FlashCardFaceType } from "../../lib/flashcard/FlashCardFace";
import { Actions } from "../actions";
import card from "../card";

describe("newly added card", () => {
    const setId = "test";
    const action = Actions.addNewCard(setId);

    it("has an id", () => {
        const newCard = card(undefined, action);
        expect(newCard.id).toBeTruthy();
    });

    it("has faces with correct ids", () => {
        const newCard = card(undefined, action);
        expect(newCard.faces.front).toBeTruthy();
        expect(newCard.faces.front.id).toBe("front");
        expect(newCard.faces.back).toBeTruthy();
        expect(newCard.faces.back.id).toBe("back");
    });

    it("has correct set id", () => {
        const newCard = card(undefined, action);
        expect(newCard.setId === setId);
    });

    it("has faces with correct set ids", () => {
        const newCard = card(undefined, action);
        expect(newCard.faces.front.setId).toBe(setId);
        expect(newCard.faces.back.setId).toBe(setId);
    });
});

describe("existing card added", () => {
    const setId = "test";
    const cardId = "test";
    const addedCard: IFlashCard = {
        setId,
        id: cardId,
        faces: {
            front: {
                id: "front",
                type: FlashCardFaceType.None,
                setId,
                cardId,
            },
            back: {
                id: "front",
                type: FlashCardFaceType.RichText,
                setId,
                cardId,
                richTextContent: ContentState.createFromText("test"),
            },
        },
    };

    it("all properties added", () => {
        expect(card(addedCard, Actions.addNewCard(setId))).toEqual(addedCard);
    });

    it("throw error on wrong face ids", () => {
        const cardErrorFront: IFlashCard = {
            ...addedCard,
            faces: {
                front: {
                    ...addedCard.faces.front,
                    id: "back",
                },
                back: {
                    ...addedCard.faces.back,
                },
            },
        };
        expect(card(cardErrorFront, Actions.addNewCard(setId))).toThrowErrorMatchingSnapshot();

        const cardErrorBack: IFlashCard = {
            ...addedCard,
            faces: {
                front: {
                    ...addedCard.faces.front,
                },
                back: {
                    ...addedCard.faces.back,
                    id: "front",
                },
            },
        };
        expect(card(cardErrorBack, Actions.addNewCard(setId))).toThrowErrorMatchingSnapshot();
    });
});

describe("swap card faces", () => {
    const setId = "test";
    const cardId = "test";
    const swappedCard: IFlashCard = {
        setId,
        id: cardId,
        faces: {
            front: {
                id: "front",
                type: FlashCardFaceType.None,
                setId,
                cardId,
            },
            back: {
                id: "front",
                type: FlashCardFaceType.RichText,
                setId,
                cardId,
                richTextContent: ContentState.createFromText("test"),
            },
        },
    };

    expect(card(swappedCard, Actions.swapCardFaces(swappedCard.setId, swappedCard.id)))
        .toEqual(
            {
                ...swappedCard,
                faces: {
                    front: {
                        ...swappedCard.faces.back,
                        id: "front",
                    },
                    back: {
                        ...swappedCard.faces.front,
                        id: "back",
                    },
                },
            },
        );
});
