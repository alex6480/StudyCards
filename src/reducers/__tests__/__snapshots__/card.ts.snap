// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`existing card added throw error on wrong face ids 1`] = `"received is not a function"`;

exports[`existing card added throw error on wrong face ids 2`] = `"received is not a function"`;
